testflinger_agent/job.py

```diff
@@ -133,85 +147,66 @@ def run_with_log(self, cmd, logfile, cwd=None):
-                # Check if there's any new data, timeout after 10s
-                data_ready = readpoll.poll(10000)
-                if data_ready:
-                    buf = process.stdout.read().decode(
-                        sys.stdout.encoding, errors="replace"
-                    )
-                    if buf:
-                        sys.stdout.write(buf)
-                        live_output_buffer += buf
-                        f.write(buf)
-                        f.flush()
-                else:
-                    if (
-                        self.phase == "test"
-                        and time.time() - buffer_timeout > output_timeout
-                    ):
-                        buf = (
-                            "\nERROR: Output timeout reached! "
-                            "({}s)\n".format(output_timeout)
-                        )
-                        live_output_buffer += buf
-                        f.write(buf)
-                        process.kill()
-                        break
-                if (
-                    self.phase != "reserve"
-                    and time.time() - start_time > global_timeout
-                ):
-                    buf = "\nERROR: Global timeout reached! ({}s)\n".format(
-                        global_timeout
-                    )
-                    live_output_buffer += buf
-                    f.write(buf)
-                    process.kill()
-                    break
-                # Don't spam the server, only flush the buffer if there
-                # is output and it's been more than 10s
-                if live_output_buffer and time.time() - buffer_timeout > 10:
-                    buffer_timeout = time.time()
-                    # Try to stream output, if we can't connect, then
-                    # keep buffer for the next pass through this
-                    if self.client.post_live_output(
-                        self.job_id, live_output_buffer
-                    ):
-                        live_output_buffer = ""
-            buf = process.stdout.read()
-            if buf:
-                buf = buf.decode(sys.stdout.encoding, errors="replace")
-                sys.stdout.write(buf)
-                live_output_buffer += buf
-                f.write(buf)
-            if live_output_buffer:
-                self.client.post_live_output(self.job_id, live_output_buffer)
-            try:
-                status = process.wait(10)  # process.returncode
-            except TimeoutError:
-                status = 99  # Default in case something goes wrong
-            return status
+                return _get_exit_code(process)
+
+            # Check for global timeout for any phase except reserve
+            if (
+                self.phase != "reserve"
+                and time.time() - start_time > global_timeout
+            ):
+                output = (
+                    f"\nERROR: Global timeout reached! ({global_timeout}s)\n"
```

Comment:
I'm wondering if it would be clearer to add the new line char in the `send_output` function, since we usually expect a log message to be written from the new line

Reply:
We're also using _send_output() to send the output from the device agent to both the logfile and to the server as it runs. We don't want to add additional newlines to that output, or else the output from running tests and provisioning will look weird.

Reply:
however I just remembered another reason for the complexity... In the previous implementation, we ensured that the output made it to the server. If it didn't, we buffered it and retried later. This simpler version no longer does that, which means there's a possibility that we drop some of the output (at least while the job is in progress). I need to think about that some more.

Reply:
I think that we can add new line when we write a log message to the log file, and not to add it when we send it to a server. Because now we also send this new line to the server